util/changes.go

```diff
@@ -0,0 +1,42 @@
+package util
+
+import (
+	"reflect"
+)
+
+// RecursivelySetUndefinedFields sets all fields that are not set in structToSet to the value of the corresponding field in structToDefaultFrom.
+// It takes two reflect.Values, the first is the struct to set the fields on, and the second is the struct to get the default values from.
+func RecursivelySetUndefinedFields(structToSet, structToDefaultFrom reflect.Value) {
+	if structToSet.Kind() == reflect.Ptr {
+		structToSet = structToSet.Elem()
+	}
+	if structToDefaultFrom.Kind() == reflect.Ptr {
+		structToDefaultFrom = structToDefaultFrom.Elem()
+	}
+
+	// Iterate through each field of the struct.
+	for i := 0; i < structToSet.NumField(); i++ {
+		// If the field we are defaulting from is undefined we can skip it.
+		if IsFieldUndefined(structToDefaultFrom.Field(i)) {
+			continue
+		}
+		branchField := structToSet.Field(i)
+
+		// If the field isn't set, use the default field.
+		// Note for pointers and maps, we consider the field undefined if the item is nil or empty length,
+		// and we don't check for subfields. This allows us to group some settings together as defined or undefined.
+		if IsFieldUndefined(branchField) {
+			reflectedField := structToDefaultFrom.Field(i)
+			branchField.Set(reflectedField)
+
+			// If the field is a struct and isn't undefined, then we check each subfield recursively.
+		} else if branchField.Kind() == reflect.Struct {
+			RecursivelySetUndefinedFields(branchField, structToDefaultFrom.Field(i))
+		} else if IsFieldPtr(branchField) {
+			// If the field is a struct and isn't undefined, then we check each subfield recursively.
```

Comment:
Here are you saying, "if the nested structure is a pointer, but is undefined, still set fields recursively?" I'm pretty sure this will break ParserProject logic, which assumes that defined pointers are themselves a defined field, even if not all inner fields are set (although I'm not sure[ if the test](https://github.com/evergreen-ci/evergreen/blob/9ebdda2fa9564930e86737e2cde6408c1ce94382/model/project_ref_test.go#L110) covers that case, may need to add to it to verify). 

If this is the case, we may need to add a boolean like "recurseOnPointers".

(Also nit that this comment shouldn't just be copied over).

Reply:
This also is reflected (no pun intended but lol) in the comment:
```
		// and we don't check for subfields. This allows us to group some settings together as defined or undefined.
```

Reply:
@khelif96 did you see this comment? I'm still concerned about this.

Reply:
I think that is guarded against on line 76 by this case in `IsFieldUndefined`
https://github.com/evergreen-ci/evergreen/blob/44ce47e9fb36162d512f29f0fd890874b699221c/util/reflect.go#L27-L28
if the branchField is a pointer and is undefined then we just defaultFrom the whole field. We don't recurse on it in that case and we set the whole struct. (Thus setting defined and undefined fields.)

We only recurse on it if the field is a struct or is a pointer to a defined struct. 

I think this test should capture that case but feel free to correct me if i'm wrong.
https://github.com/evergreen-ci/evergreen/pull/5625/files#diff-c2b5cccbb3c39a7a30e5046a24c02a28a7fe993a16dbbcfd283c80390491fad1R71-R92

Reply:
Oh yeah that actually looks good; it seems we only don't recurse if it's a struct (and not a pointer) and that's maintained.